pkg/apis/kops/v1alpha2/instancegroup.go

```diff
@@ -159,6 +159,9 @@ type WarmPoolSpec struct {
 	// (unless the resulting number is smaller than MinSize).
 	// The default is the instance group's MaxSize.
 	MaxSize *int64 `json:"maxSize,omitempty"`
+	// EnableLifecyleHook determines if an ASG lifecycle hook will be added ensuring that nodeup runs to completion.
+	// Note that the metadata API must be protected from arbitrary Pods when this is enabled.
+	EnableLifecyleHook bool `json:"enableLifecycleHook,omitempty"`
```

Comment:
I would like to question the need for this setting, or at least its default.

What is the risk of enabling this when the metadata API is not protected from arbitrary pods, over and above the risk of having the metadata API be not protected even without enabling this?

Reply:
Does https://kops.sigs.k8s.io/instance_groups/#lifecycle-hook answer the question?

Reply:
The documentation states that the risk is that a workload pod could complete the lifecycle hook with the `ABANDONED` result, which would prevent subsequently created instances from joining the cluster (it is not clear to me how that would be). Would the pod be able to abandon the lifecycle hook even though the instance wouldn't be warming? Would the pod not be able to achieve the same result by abandoning some other lifecycle hook?

Reply:
The lifecycle hook is used for any launching instance, both those going into the warm pool and those going directly into the ASG. The risk is introduced both because we define the launching hook and because we give the node IAM role the ability to complete them.

Reply:
Can the lifecycle hook be abandoned after it has been completed? Could we not complete that lifecycle hook in nodeup even when the instance isn't warming?

Reply:
I'm trying to figure out how this is particularly worse than giving pods access to the rest of the instance role's permissions.

We happen to mitigate this by having a post-nodeup hook which asks a lambda to change the instance role to one with only DescribeInstance.

Reply:
We complete the hook in all cases, if not the instance would be stuck `pending:wait` until it is terminated because it fails to complete. It can't be abandoned after being completed with `CONTINUE`, but that happens last in nodeup. So plenty of time for a pod to iterate across all instance and try to abandon the lifecycle hook action.

But I don't think there is much case for running with max-hop-limit > 1 on regular nodes. So we may just set that as default. 

We tried to set that as default on all nodes, I believe, but that caused CP nodes to fail because pods there do need to talk to the metadata API.

I.e if we set max-hop-limit = 1 by default, we can flip the default on for the lifecycle hooks as well.

Reply:
We could start with max-hop-limit=1 for worker nodes to start with. Once we have IRSA in we could also make that the default for control plane nodes when IRSA is enabled.

We should file a ticket with AWS asking for the ability to restrict the IAM permission to same instance.

Reply:
Yeah. That makes sense.